src/output/plugins/visualization/SoundAnalysis.cxx

```diff
@@ -0,0 +1,394 @@
+								   NULL,
+								   1,			  // input stride
+								   dist,		  // distance between successive arrays (indexes, not bytes)
+								   out.get(),	  // output buffer-- overwritten on each execution
+								   NULL,
+								   1,			  // output stride
+								   odist,		  // distance between successive arrays (indexes, not bytes)
+								   FFTW_ESTIMATE);// should probably be zero (to select a more exhaustive
+												  // search), but out of an abundance of caution, tell
+												  // FFTW to pick a plan quickly
+	if (NULL == plan) {
+		throw FmtRuntimeError("Failed to generate an FFTW plan: num_samp={},num_chan={}",
+							  num_samples, num_channels);
+	}
+
+	freq_lo = params.GetLoCutoff();
+
+	float samples_per_sec = (float) audio_format.GetSampleRate();
+	float ns = (float) num_samples;
+	// The highest frequency we can represent will be
+	float max_freq = (ns - 1.0f) * samples_per_sec / ns;
+	if (max_freq < params.GetHiCutoff()) {
+		FmtWarning(d_sound_analysis, "Clamping max frequency from {} to {}", freq_hi, max_freq);
+		freq_hi = max_freq;
+	} else {
+		freq_hi = params.GetHiCutoff();
+	}
+
+	idx_lo = (size_t) floor(freq_lo * (float) num_samples / samples_per_sec );
+	idx_hi = (size_t) ceil(freq_hi * (float)num_samples / samples_per_sec);
+
+	float num_octaves = log(freq_hi/freq_lo) / 0.69314718f;
+
+	float freq_mids	 = freq_lo * pow(2.0f, num_octaves / 3.0f);
+	float freq_trebs = freq_lo * pow(2.0f, 2.0f * num_octaves / 3.0f);
+
+	idx_mids  = ns *  freq_mids / samples_per_sec;
+	idx_trebs = ns * freq_trebs / samples_per_sec;
+}
+
+bool
+Visualization::SoundAnalysis::Update(SoundInfoCache::Time t) noexcept
+{
+	FmtDebug(d_sound_analysis, "SoundAnalysis::update(tid: {}), time {}us, # samp: {}, "
+			 "buffer size: {}", gettid(), t.count(), num_samples, pcache->Size());
+
+	if (!pcache->GetByTime(num_samples, t, buf.get(), cbbuf)) {
+		FmtWarning(d_sound_analysis, "Failed to get samples by time for sound analysis "
+				   "({} samples requested, at time {}us for buf size {}).", num_samples, 
+				   t.count()/1000, cbbuf);
+		return false;
+	}
+
+	/* Copy the raw PCM data from `buf` into `in`. I hate this, but we need to
+	 * convert the input data from `uint16_t` (or whatever) to `float`
+	 * regardless. We could, of course, do the conversion when the PCM data is
+	 * added to the cache, but since I anticipate processing far fewer samples
+	 * than I store, I expect this to be more efficient (both in terms of time &
+	 * space).
+	 * 
+	 * Since we have to do the copy anyway, let's convert from interleaved to
+	 * sequential (i.e. all samples for the first channel laid-out contiguously,
+	 * followed by all from the second, and so forth). */
+	typedef IntegerToFloatSampleConvert<SampleFormat::S8> s8_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S16> s16_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S32> s32_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S24_P32> s24_p32_t;
```

Comment:
Naming convention for types in MPD is CamelCase, not foo_bar_t. Also please use `using` instead of the deprecated `typedef` keyword.